File: week_11/day_3/enumeration_lab_start_point/enumeration.js
var Enumeration = function() {}

Enumeration.prototype.find = function(array, callback) {
  for (element of array) {
    if (callback(element)){
      return element;
    }
  }
}

Enumeration.prototype.map = function (array, callback) {
  const newArray = [];
  for (element of array) {
    newArray.push(callback(element));
  }
  return newArray;
};

Enumeration.prototype.filter = function (array, callback) {
  const newArray = [];
  for (element of array) {
    if (callback(element)){
      newArray.push(element);
    }
  }
  return newArray;
};

Enumeration.prototype.some = function (array, callback) {
  for (element of array) {
    if (callback(element)){
      return true;
    }
  }
  return false;
};

Enumeration.prototype.every = function (array, callback) {
  for (element of array) {
    if (!callback(element)){
      return false;
    }
  }
  return true;
};

Enumeration.prototype.reduce = function (array, callback) {
  let total = 0;
  for (element of array) {
    total = callback(total, element);
  }
  return total;
};

module.exports = Enumeration;

File: week_02/day_4/enumeration/enumeration.rb
chickens = ['Margaret', 'Hetty', 'Henrietta', 'Audrey', 'Mabel']

# ENUMERATION

# #using for loops
# for chicken in chickens
#   p chicken
# end


# # EACH - same as a for loop
# chickens.each {|chicken| p chicken}
#
# #multiple lines -> use do and end
# chickens.each do |chicken|
#   shouting_chicken = chicken.upcase()
#   p shouting_chicken
# end

# # MAP - takes elements and returns them
# name_lengths = []
# for chicken in chickens
#   name_lengths << chicken.length()
# end
# p name_lengths
#
# name_lengths = chickens.map {|chicken| chicken.length()}
# p name_lengths

# # EACH WITH INDEX - takes element and index and returns both
# chickens.each_with_index {|chicken, index| p "#{chicken} is at index #{index}"}
#
# # FIND - finds first in array that matches condition
# p chickens.find {|chicken| chicken[0] == "H"}
#
# # FINDALL - finds all in array that match condition, gives an array back
# # SELECT - alt to FIND ALL, also gives an array back
# p chickens.find_all {|chicken| chicken[0] == "H"}
# p chickens.select {|chicken| chicken[0] == "H"}
#
# # REDUCE - takes multiple variables from an array and combines them
# p chickens.reduce {|list, chicken| list + ", " + chicken}
# # can also be used with numbers (integers or floats)
# numbers = [1,2,3,4,5]
# p numbers.reduce() {|sum, number| sum + number}
# but note that if the first item is not a number you need an initial value also
# drinks = [
#     {name: "Pepsi", quantity: 10, price: 1},
#     {name: "Budweiser", quantity: 10, price: 3}
# ]
# p drinks.inject(0) {|total, drink| total + (drink[:quantity] * drink[:price]) }
